src/lightning/data/datasets/iterable.py

```diff
@@ -0,0 +1,447 @@
+            but may result in undeterministic behavior.
+
+        Note:
+            On resume from a state-dict, we always skip currently started chunks as these would make the data-order
+            impossible to determine with sharding. Upon resuming from a point, where a new chunk would be started
+            anyways, nothing is skipped.
+
+        Note:
+            Order of data is only guaranteed when resuming with the same distributed settings and the same number of
+            workers. Everything else leads to different sharding and therefore results in different data order.
+    """
+
+    # TODO: backend docstring
+    def __init__(
+        self,
+        chunks: Sequence[Any],
+        chunk_size: int = 1,
+        num_parallel_chunks: int = 1,
+        env: Optional[Environment] = None,
+        shuffle: bool = False,
+        seed: Optional[int] = None,
+        wrap: bool = False,
+        lazy_shuffle: bool = False,
+        backend: str = "local",
+    ):
+        _StatefulIterableDataset.__init__(self)
+        _Dataset.__init__(self, backend=backend)
+
+        chunks = [_Chunk(c, chunk_size=chunk_size) for c in chunks]
+        if env is None:
+            # must detect distributed env here since distributed is not initialized in worker processes with ddp spawn
+            # can't detect worker env here since workers not yet initialized
+            env = Environment(DistributedEnv.detect(), None)
+        self._env = env
+        self._shuffle = shuffle
+        self._lazy_shuffle = lazy_shuffle
+
+        # prepare shuffling
+        if shuffle:
+            generator = torch.Generator()
+            if seed is not None:
+                generator = generator.manual_seed(seed)
+        else:
+            generator = None
+
+        self._seed = seed
+        self._generator = generator
+        self._initial_generator_state = self._generator.get_state() if self._generator is not None else None
+
+        self._num_parallel_chunks = num_parallel_chunks
+        self._chunks = chunks
+        self._original_chunks = chunks
+
+        self._chunk_size = chunk_size
+        self._local_chunks: List[_Chunk] = []
+        self._wrap = wrap
+
+        self._start_index_chunk = 0
+        self._start_index_sample = 0
+        self._curr_chunk_index = 0
+        self._curr_sample_index = 0
+
+        self._curr_loaded_chunks: List[_Chunk] = []
+        self._curr_loaded_num_samples = 0
+
+    def prepare_chunk(self, chunk: Any) -> None:
+        """Prepares a single chunk before it is actually loaded. This could e.g. download the actual file from s3.
+
+        Args:
+            chunk: the chunk data to prepare.
+        """
+
+    @abstractmethod
+    def load_chunk(self, chunk: Any) -> Any:
+        """Loads a single chunk into memory. This could e.g. mean loading the file that has previously been
+        downloaded from s3.
```

Comment:
Since the user needs to implement this hook, maybe an imperative style docstring could be better? E.g., "Implement this to load a single chunk into memory ...". Same suggestion below